File: src/contrast.rs
use super::data::Matrix;

use arrayfire::*;

pub enum ContrastFunctionId {
    LOGCOSH = 0,
    KURTOSIS = 1,
    EXP = 2,
}

pub type ContrastFunc = fn(&Matrix, f64) -> (Matrix, Matrix);

const CONTRAST_FUNCTIONS: [ContrastFunc; 3] = [logcosh, kurtosis, exponential];

pub fn get_contrast_function(fid: ContrastFunctionId) -> ContrastFunc {
    CONTRAST_FUNCTIONS[fid as usize]
}

fn logcosh(mat: &Matrix, alpha: f64) -> (Matrix, Matrix) {
    let factored_mat = mul(mat, &alpha, true);

    // 1/alpha * log(cosh(alph * u))
    let g = log(&cosh(&factored_mat));
    
    let g = mul(&(1.0 / alpha), &g, true);

    // np.tanh(alph * u)
    let dg = tanh(&factored_mat);

    (g, dg)
}

fn kurtosis(mat: &Matrix, alpha: f64) -> (Matrix, Matrix) {
    (
        mul(&(1.0 / alpha), &pow(mat, &alpha, true), true),
        pow(mat, &(alpha - 1.0), true),
    )
}

fn exponential(mat: &Matrix, alpha: f64) -> (Matrix, Matrix) {
    let e = {
        let mut temp = pow(mat, &alpha, true);
        temp = sub(&0, &temp, true);

        let coef = -(1.0 / alpha);

        mul(&coef, &temp, true)
    };

    let g = mul(mat, &e, false);

    // (1 - power(u, 2)) * exp
    let dg = {
        let mut temp = pow(mat, &alpha, true);
        temp = mul(&temp, &e, false);
        sub(&1, &temp, true)
    };

    (g, dg)
}

File: src/contrast.cpp
#include "../include/contrast.h"
#include <Eigen/Dense>

using namespace Eigen;

namespace contrast
{

    static ContrastFunction CONTRAST_FUNCTIONS[2] = {
        &contrast::logcosh,
        &contrast::exp
    };

    ContrastFunction get_contrast_function(ContrastFunctionId cfid)
    {
        return CONTRAST_FUNCTIONS[cfid];
    }

    MatPair logcosh(const Eigen::MatrixXd &mat, double alpha)
    {
        // todo
        MatrixXd factored_mat = alpha * mat;

        MatrixXd dg  = factored_mat.array().tanh();

        MatrixXd d2g = 1 - dg.array().pow(2);

        return MatPair(dg, alpha * d2g);
    }

    MatPair exp(const Eigen::MatrixXd &mat, double alpha)
    {
        MatrixXd dg  =  mat.array().pow(2) * 1.0 / alpha;

        MatrixXd d2g =  mat.array().pow(alpha - 1.0);

        return MatPair(dg, d2g);
    }

} // namespace contrast